fs/unpacker.go

```diff
@@ -46,56 +47,177 @@ type Archive interface {
 }
 
 type layerArchive struct {
-	verifier         *layerVerifier
+	compressed       *asyncVerifier
+	uncompressed     *asyncVerifier
 	decompressStream socicompression.DecompressStream
+	bufferPool       *bufferPool
 }
 
-type layerVerifier struct {
-	compressed   digest.Verifier
-	uncompressed digest.Verifier
+type asyncVerifier struct {
+	v       digest.Verifier
+	waitCh  chan struct{}
+	started bool
 }
 
-func NewLayerArchive(compressedVerifier, uncompressedVerifier digest.Verifier, decompressStream socicompression.DecompressStream) Archive {
+func newAsyncVerifier(v digest.Verifier) *asyncVerifier {
+	return &asyncVerifier{
+		v:       v,
+		waitCh:  make(chan struct{}, 1),
+		started: false,
+	}
+}
+
+func (av *asyncVerifier) AsyncVerify(reader io.ReadCloser) {
+	if av == nil || av.v == nil {
+		reader.Close()
+		return
+	}
+	go func() {
+		io.Copy(av.v, reader)
+		close(av.waitCh)
+		reader.Close()
+	}()
+	av.started = true
+}
+
+func (av *asyncVerifier) Verified(ctx context.Context) bool {
+	if av == nil || av.v == nil {
+		return true
+	}
+	if !av.started {
+		return false
```

Comment:
Is there ever a race condition in the code path where it has not started at this point? I don't think so but just something to think about

Reply:
Right. not in current implementation. since the async operation is always started by the fetch much earlier.